### DISCUSSION TITLE
Proper usage of 'engine' in multithreaded applications

### DISCUSSION THREAD

**Discussion by tinez** [sqlalchemy/sqlalchemy#11740] [Usage Questions]
Hi, first of all I'd like to thank you for this library, I think it's awesome!

I'm having problems with a multithreaded application based on `Python 3.12.4` and `SQLAlchemy-2.0.32` with `sqlite+pysqlite` underneath. I've boiled it down to a simple script that updates/retrieves data in the DB in a loop. When the `engine` is created in the main thread and passed to the workers, the script reports multiple different errors when executing, e.g. `sqlalchemy.exc.MultipleResultsFound: Multiple rows were found when one or none was required`, `sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked`. When the `engine` is created by each worker separately, there are no problems reported at all.

I went over both sqlite and SqlAlchemy docs and my understanding is that:
* sqlite by default is opened in serialized mode, which should work well with multiple threads
* the 'engine' class itself is thread-safe, but 'Session' and 'Connection' objects are not

What I tried to make it work (other than creating 'engine' per thread):
* adding [these](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#serializable-isolation-savepoints-transactional-ddl) event handlers
* placing the db on different filesystems

Am I doing things right? Did I miss anything? Any help would be appreciated.

Here's the script I'm using for testing:
```python
import dataclasses
import random
import threading
from typing import Any

from sqlalchemy import (
    Column,
    Engine,
    MetaData,
    Table,
    Text,
    create_engine,
    insert,
    select,
    update,
)

CYCLE_COUNT = 1000
DB_PATH = "/tmp/my.db"
POOL_SIZE = 10
THREAD_COUNT = 8


@dataclasses.dataclass(frozen=True)
class Validator:
    uid: str
    resource_id: str
    status: str


TEST_DATA = [Validator(uid=f"uid-{i}", resource_id=f"rid-{i}", status="") for i in range(5)]


class ValidatorsDao:
    _meta = MetaData()
    _validator_runs = Table(
        "Validators",
        _meta,
        Column("uid", Text, primary_key=True),
        Column("resource_id", Text, nullable=False),
        Column("status", Text, nullable=False),
    )

    def __init__(self, engine: Engine):
        self._engine = engine
        self._initialized = False

    def initialize(self) -> None:
        if not self._initialized:
            self._meta.create_all(self._engine)
            self._initialized = True

    def add(self, validator: Validator) -> None:
        with self._engine.begin() as connection:
            connection.execute(
                insert(self._validator_runs), [dataclasses.asdict(validator)]
            )

    def get_by_resource_id(self, resource_id: str) -> Validator | None:
        with self._engine.begin() as connection:
            result = connection.execute(
                select(self._validator_runs).where(
                    self._validator_runs.c.resource_id == resource_id
                )
            )
        row = result.one_or_none()
        if row is None:
            return None
        return Validator(**row._mapping)

    def update(self, old: Validator, new: Validator) -> None:
        with self._engine.begin() as connection:
            connection.execute(
                update(self._validator_runs)
                .where(self._validator_runs.c.uid == old.uid)
                .values(dataclasses.asdict(new))
            )


def _random_resource_id() -> str:
    return random.choice([validator.resource_id for validator in TEST_DATA])


def _create_engine() -> Any:
    uri = f"sqlite+pysqlite:///{DB_PATH}"
    return create_engine(url=uri, pool_size=POOL_SIZE)


def worker(engine):
    print("Worker here")

    ############################################
    # UNCOMMENT THE LINE BELOW TO MAKE IT WORK #
    ############################################
    # engine = _create_engine()
    dao = ValidatorsDao(engine)
    counter = 1

    for _ in range(CYCLE_COUNT):
        if random.random() < 0.5:
            # modify validator
            validator = dao.get_by_resource_id(_random_resource_id())

            assert validator is not None

            new = dataclasses.replace(validator, status=f"{threading.current_thread().name}-{counter}")
            dao.update(validator, new)
            counter += 1
        else:
            # retrieve validator
            resource_id = _random_resource_id()
            validator = dao.get_by_resource_id(resource_id)

    print("Worker done")


if __name__ == "__main__":
    engine = _create_engine()
    dao = ValidatorsDao(engine)
    dao.initialize()

    print("Populating db...")
    for v in TEST_DATA:
        dao.add(v)

    print("Spawning threads...")
    workers = []
    for _ in range(THREAD_COUNT):
        workers.append(threading.Thread(target=worker, args=[engine]))
        workers[-1].start()

    print("Waiting for workers")
    for w in workers:
        w.join()

    print("Done")
```

**Comment by CaselIT** [comment]
Hi,

Haven't had time to try this yes, but by a quick look it seems ok. Have you tried also with other databases? Like postgresql?

**Reply by tinez** [reply]
Hi, thanks for your reply!

I haven't tried other DBs yet. We might migrate our application to postgresql in the future, but for this moment we'd prefer to stick to sqlite. My colleague claimed that the code worked when using sqlite directly, but it's possible it did because we followed a similar "resources per thread" approach like the one that fixes the script above.

I'll take any other hints happily. I'll also try rewrite the script to use sqlite directly to see if our claims were true.

**Reply by tinez** [reply]
Ok, I just noticed something... in `get_by_resource_id`, when we pull the `row` out of the `result`, the `connection` context manager already exited... Do I understand correctly, that the rows are pulled from the results (and hence, the connection) lazily and the connection may be returned to the pool and reused by another thread before we extract the data in that scenario? I indented that part that pulls the results to live as long as the connection and the errors magically disappeared :smile:

**Reply by CaselIT** [reply]
That seems like the case.

**Reply by AvlWx2014** [reply]
Sorry to revive this again after a few weeks, but interestingly we just encountered this issue again today despite updating our database layer code to fix the issue @tinez described above w.r.t. iterating over a cursor lazily. 

My next suspicion is something like this:

```python
        with self._engine.connect() as connection:
            result = connection.execute(
                select(self._validator_runs).where(
                    self._validator_runs.c.run_id == str(run_id)
                )
            )
            return map(_from_row, result.all())
```

The built-in `map` returns an iterator, which means calling `_from_row` with each `row` in `result.all()` is done lazily, but as far as I can tell `result.all()` itself is eager, so it should return the entire list of rows before the `map` iterator is constructed. Am I on the right track that `result.all()` eagerly exhausts the cursor?

**Reply by CaselIT** [reply]
It does, yes. All returns a list that's fully populated

**Reply by AvlWx2014** [reply]
Awesome, thanks! I was kind of hoping you'd say "no it's lazy!" because then I'd have some idea of where to look next lol but since I was on the right track I'll keep looking. Thanks again!

**Reply by AvlWx2014** [reply]
What's most curious about this is that the error seems to arise in a single-threaded subsystem where we have a control loop running on a dedicated thread, polling for status updates and updating the database accordingly. It seems like somehow a single thread is locking itself out of the database :thinking:.

**Reply by zzzeek** [reply]
you dont need threads to create deadlocks, any reentrant call to a lock that is not itself reentrant will deadlock and this is probably the more common cause of development-time deadlocks (I say development time because they tend to be deterministic and are identified before code goes live)
